src/server/pfs/server/trigger.go

```diff
@@ -7,50 +7,82 @@ import (
 	"google.golang.org/protobuf/proto"
 
 	"github.com/pachyderm/pachyderm/v2/src/internal/ancestry"
+	col "github.com/pachyderm/pachyderm/v2/src/internal/collection"
 	"github.com/pachyderm/pachyderm/v2/src/internal/cronutil"
 	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pfsdb"
 	"github.com/pachyderm/pachyderm/v2/src/internal/transactionenv/txncontext"
 	"github.com/pachyderm/pachyderm/v2/src/pfs"
 )
 
 // triggerCommit is called when a commit is finished, it updates branches in
 // the repo if they trigger on the change
-func (d *driver) triggerCommit(
-	txnCtx *txncontext.TransactionContext,
-	commit *pfs.Commit,
-) error {
-	repoInfo := &pfs.RepoInfo{}
-	if err := d.repos.ReadWrite(txnCtx.SqlTx).Get(commit.Repo, repoInfo); err != nil {
-		return errors.EnsureStack(err)
-	}
-	newHead := &pfs.CommitInfo{}
-	if err := d.commits.ReadWrite(txnCtx.SqlTx).Get(commit, newHead); err != nil {
-		return errors.EnsureStack(err)
-	}
-	for _, b := range repoInfo.Branches {
-		bi := &pfs.BranchInfo{}
-		if err := d.branches.ReadWrite(txnCtx.SqlTx).Get(b, bi); err != nil {
-			return errors.EnsureStack(err)
+func (d *driver) triggerCommit(txnCtx *txncontext.TransactionContext, commitInfo *pfs.CommitInfo) error {
+	branchInfos := make(map[string]*pfs.BranchInfo)
+	branchInfo := &pfs.BranchInfo{}
+	if err := d.branches.ReadWrite(txnCtx.SqlTx).GetByIndex(pfsdb.BranchesRepoIndex, pfsdb.RepoKey(commitInfo.Commit.Repo), branchInfo, col.DefaultOptions(), func(_ string) error {
+		branchInfos[pfsdb.BranchKey(branchInfo.Branch)] = proto.Clone(branchInfo).(*pfs.BranchInfo)
+		return nil
+	}); err != nil {
+		return err
+	}
+	// Recursively check / fire trigger chains.
+	newHeads := make(map[string]*pfs.CommitInfo)
+	newHeads[pfsdb.BranchKey(commitInfo.Commit.Branch)] = commitInfo
+	var triggerBranch func(*pfs.BranchInfo) (*pfs.CommitInfo, error)
+	triggerBranch = func(bi *pfs.BranchInfo) (*pfs.CommitInfo, error) {
+		branchKey := pfsdb.BranchKey(bi.Branch)
+		head, ok := newHeads[branchKey]
+		if ok {
+			return head, nil
+		}
+		newHeads[branchKey] = nil
+		if bi.Trigger == nil || bi.Trigger.CronSpec != "" {
+			return nil, nil
+		}
+		// Recurse through the trigger chain, checking / firing earlier triggers first.
+		triggerBranchKey := pfsdb.BranchKey(bi.Branch.Repo.NewBranch(bi.Trigger.Branch))
+		triggerBranchInfo, ok := branchInfos[triggerBranchKey]
+		// TODO: We probably shouldn't allow the creation of a trigger on a nonexistent branch.
+		if !ok {
+			return nil, nil
+		}
+		newHead, err := triggerBranch(triggerBranchInfo)
```

Comment:
I dont think I understand what's going on here, I was expecting this to be in a for loop.   Is this going upstream?